dashboard/app/models/plc/user_course_enrollment.rb

```diff
@@ -0,0 +1,46 @@
+# == Schema Information
+#
+# Table name: plc_user_course_enrollments
+#
+#  id            :integer          not null, primary key
+#  status        :string(255)
+#  plc_course_id :integer
+#  user_id       :integer
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#
+# Indexes
+#
+#  index_plc_user_course_enrollments_on_plc_course_id  (plc_course_id)
+#  index_plc_user_course_enrollments_on_user_id        (user_id)
+#
+
+class Plc::UserCourseEnrollment < ActiveRecord::Base
+  belongs_to :plc_course, class_name: '::Plc::Course'
+  belongs_to :user, class_name: 'User'
+  has_many :module_assignments, class_name: '::Plc::EnrollmentModuleAssignment', foreign_key: 'plc_user_course_enrollment_id', dependent: :destroy
+  has_many :task_assignments, through: :module_assignments, class_name: '::Plc::EnrollmentTaskAssignment', dependent: :destroy
+
+  validates :user, presence: true
+  validates :plc_course, presence: true
+
+  def complete_course
+    self.status = :completed
+    self.save!
+  end
+
+  def enroll_user_in_course_with_learning_modules learning_modules
```

Comment:
Under normal circumstances, task assignment should not fail. But yes - module assignment and task assignment should be in a transaction so it can be rolled back.